tools/structwrite/structwrite.go

```diff
@@ -0,0 +1,231 @@
+package structwrite
+
+import (
+	"go/ast"
+	"go/token"
+	"go/types"
+	"strings"
+
+	"github.com/golangci/plugin-module-register/register"
+	"golang.org/x/tools/go/analysis"
+)
+
+func init() {
+	register.Plugin("structwrite", New)
+}
+
+// Settings defines the configuration schema for the plugin.
+type Settings struct {
+	// Structs is a list of fully-qualified struct type names for which immutability is enforced.
+	// TODO: define in godoc instead?
+	Structs []string `json:"structs"`
+
+	// ConstructorRegex is a regex pattern (optional) to identify allowed constructor function names.
+	// TODO: unused - current just use New.*
+	ConstructorRegex string `json:"constructorRegex"`
+}
+
+// PluginStructWrite implements the LinterPlugin interface for the structwrite linter.
+type PluginStructWrite struct {
+	structs map[string]bool
+}
+
+// New creates a new instance of the PluginStructWrite plugin.
+func New(cfg any) (register.LinterPlugin, error) {
+	s, err := register.DecodeSettings[Settings](cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	structMap := make(map[string]bool)
+	for _, name := range s.Structs {
+		structMap[name] = true
+	}
+
+	return &PluginStructWrite{structs: structMap}, nil
+}
+
+func (p *PluginStructWrite) BuildAnalyzers() ([]*analysis.Analyzer, error) {
+	a := &analysis.Analyzer{
+		Name: "structwrite",
+		Doc:  "flags writes to specified struct fields or construction of structs outside constructor functions",
+		Run:  p.run,
+	}
+	return []*analysis.Analyzer{a}, nil
+}
+
+func (p *PluginStructWrite) GetLoadMode() string {
+	return register.LoadModeTypesInfo
+}
+
+// run is the main analysis function.
+func (p *PluginStructWrite) run(pass *analysis.Pass) (interface{}, error) {
+	for _, file := range pass.Files {
+		ast.Inspect(file, func(n ast.Node) bool {
+			switch node := n.(type) {
+			case *ast.AssignStmt:
+				p.handleAssignStmt(node, pass, file)
```

Comment:
better to add some example of the code that we are matching here

Reply:
Added examples [here](https://github.com/onflow/flow-go/pull/7310/files#diff-df50c4899b1555225511613798c685d9cf2bd8375b65b5204294efba8186c54dR144-R166) and [here](https://github.com/onflow/flow-go/pull/7310/files#diff-df50c4899b1555225511613798c685d9cf2bd8375b65b5204294efba8186c54dR189-R207)